### DISCUSSION TITLE
Headless mode for browser or SSR

### DISCUSSION THREAD

**Discussion by linonetwo** [TiddlyWiki/TiddlyWiki5#6590] [Ideas]
In https://github.com/Jermolene/TiddlyWiki5/issues/6565#issuecomment-1087212612 :

>> And maybe even refactor the core parts of tw, so they can be used in a "headless" way, for example, a browser extension that renders readme.tid in the Github. This will make TiddlyWiki not "zero dependencies" anymore, so may not be a good idea.

> I'm not sure what you mean – TW5 can already be run headless under Node.js?

Let's discuss here, I actually mean render `*.tid` file anywhere in the network, for example, https://github.com/Jermolene/TiddlyWiki5/tree/master/plugins/tiddlywiki/codemirror

<img width="1619" alt="截屏2022-04-04 16 00 38" src="https://user-images.githubusercontent.com/3746270/161500090-befc16be-005a-4ade-a55d-bf06a3d53861.png">

I hope we can have this tid file rendered correctly. I think we can make a browser extension or userscript, if TiddlyWiki npm package can get rid of the node part and work completely in the browser.

This requires npm package exports a pure function, that have a tid file string as input, and HTML as output, this is what I mean by Headless mode, (but is for render 3rd party content...) like in https://github.com/Jermolene/TiddlyWiki5/discussions/6141#discussioncomment-1532861

### Server side rendering

Or maybe I can write a plugin that provides a server route `/rendering` that receives any string send from a browser extension or userscript, and render this string in the nodejs server, then send back the result...

There may be some security issues, but I think I can manage to create a sandbox before calling any API from `$tw.wiki`...

**Comment by Jermolene** [comment]
Hi @linonetwo a challenge here is transclusions of other tiddlers, or macro calls. They all require an entire wiki, not the single tiddler on its own. So perhaps your browser extension would need to walk up the tree to load the other tiddlers from the repo before trying to render the target tiddler.

**Reply by linonetwo** [reply]
sometimes we just want to render a standalong wikitext, without connection to other tiddlers. Like in [twMat - “Discussion examples” converted into tiddlers?](https://talk.tiddlywiki.org/t/discussion-examples-converted-into-tiddlers/2843)

If there is a single js + CSS version of tw (not using nodejs features to load core files), this can be achieved. I think the js bundled by `external js` is a way to get it:

```
"externaljs": [
      "--rendertiddler",
      "$:/core/templates/tiddlywiki5.js",
      "tiddlywiki5.js",
      "text/plain"
    ]
```

but currently seems it will modify the whole page, instead of just providing `$tw` API to the global?

**Reply by pmario** [reply]
The external core, still needs quite a bit of js-code to be started, but it should be possible to create a browser extension, that may detect a .tid extension and converts it to HTML. ... BUT ... I don't see a real advantage here. 

If you drag&drop the .tid file to any TiddlyWiki app, you can import it and it will be rendered. So why go the way over a browser AddOn, that needs to be maintained separately.

**Reply by linonetwo** [reply]
@pmario You can see that forum link, it will be useful to render codeblock in the forum (or even on the twitter), or any readme.tid in the github.

**Reply by Jermolene** [reply]
> but currently seems it will modify the whole page, instead of just providing `$tw` API to the global?

I quite agree that it should be possible to use the TW core in the browser without automatically rendering the page etc. I thought we had a ticket for this issue, but if not it would be helpful to make one.

**Reply by linonetwo** [reply]
@pmario https://github.com/tiddly-gittly/watch-fs/issues/4 The most useful usecase is rendering Readme.tid in github.
